crates/doc/src/inference.rs

```diff
@@ -1484,6 +1486,201 @@ impl Shape {
     }
 }
 
+const MAX_ROOT_FIELDS: usize = 750;
+const MAX_NESTED_FIELDS: usize = 200;
+
+impl Shape {
+    // Widen a Shape to make the provided AsNode value fit.
+    // Returns a hint if some locations might exceed their maximum allowable size.
+    // NOTE: If a particular location defines `additionalProperties` as a subschema, don't
+    // add the field from `AsNode`, instead jump straight to squashing the the field into `additionalProperties`
+    fn widen<'n, N>(&mut self, node: &'n N) -> bool
+    where
+        N: AsNode,
+    {
+        self.widen_inner(node, Location::Root)
+    }
+
+    fn widen_inner<'n, N>(&mut self, node: &'n N, loc: Location) -> bool
+    where
+        N: AsNode,
+    {
+        use crate::{Field, Fields};
+
+        match node.as_node() {
+            crate::Node::Object(obj) => {
+                self.type_ = self.type_ | types::OBJECT;
+
+                match self.object.additional {
+                    Some(_) => {}
+                    None => {
+                        self.object.additional = Some(Box::new(Shape {
+                            type_: types::INVALID,
+                            ..Default::default()
+                        }))
+                    }
+                }
+
+                // If a particular location defines `additionalProperties` as a subschema
+                // don't add the field from `AsNode`, instead jump straight to squashing
+                let hint = match self.object.additional.as_mut() {
+                    // We only want to squash if your `additionalProperties`
+                    // is set to something other than `false`.
+                    Some(addl) if !addl.type_.eq(&types::INVALID) => {
+                        obj.iter().enumerate().fold(false, |accum, (idx, node)| {
+                            accum || addl.widen_inner(node.value(), loc.push_item(idx))
+                        })
+                    }
+                    _ => {
+                        let (hint, new_obj_shape) = itertools::merge_join_by(
+                            self.object.properties.iter_mut(),
+                            obj.iter(),
+                            |prop, field| prop.name.cmp(&field.property().to_string()),
+                        )
+                        .fold(
```

Comment:
How about, instead of this fold, you instead `filter_map()` and `collect()` a vector of entirely-new properties which should be _added_ to `object.properties` after the iteration completes?

This does a couple things, the biggest of which is that it avoids _any_ allocation in the common case where no new properties are added, because this vector is empty:

* EoB::Left would mark !required in-place and return None
* EoB::Right would return Some((property, property-shape))
* EoB::Both would widen the child in-place and return None.

Collect the iteration result into a Vec, which releases the mutable borrow, and then `extend()` the LHS map with it's `into_iter()`. If there are no new properties, this is zero-cost.

Reply:
This is a great tip, it's much cleaner. Thanks!